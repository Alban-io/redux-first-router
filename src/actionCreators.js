export const INIT = '@@address-bar/INIT' //used internally by middleware and then action is re-written
export const NOT_FOUND = '@@address-bar/NOT_FOUND'


/** EXPORTED INIT ACTION CREATORS: 
 *  for use when ready key is not provided and 
 *  manual control of initialization is desired
*/

export function init(pathname) {
  return {
    type: INIT,
    payload: {
      pathname,
    }
  }
}

export function initThunk() {
  return (dispatch, getState) => {
    let {pathname} = getState().location
    return dispatch(init(pathname))
  }
}


/** PRIMARY/SOLE NAVIGATION ACTION CREATOR 
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'