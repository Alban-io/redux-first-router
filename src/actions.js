export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
