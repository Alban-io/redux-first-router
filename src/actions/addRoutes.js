// @flow

import { ADD_ROUTES } from '../types'
import type { RoutesMap } from '../flow-types'

export default (routes: RoutesMap, formatRoute: ?Function) => ({
  type: ADD_ROUTES,
  payload: { routes, formatRoute }
})

// NOTE: see `src/utils/formatRoutes.js` for implemenation of corresponding pathlessRoute
