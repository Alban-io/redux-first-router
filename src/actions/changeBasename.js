// @flow
import type { Action } from '../flow-types'

export default (basename: string, action: ?Action) => {
  if (!action) {
    return ({ initialLocation }) => {
      const { type, params, query, state, hash } = initialLocation
      return { type, params, query, state, hash, basename }
    }
  }

  return { ...action, basename }
}
