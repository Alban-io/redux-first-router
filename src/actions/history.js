// @flow

export const push = (path: string, state: ?Object) => ({ history }) =>
  history.push(path, state, false)

export const replace = (path: string, state: ?Object) => ({ history }) =>
  history.replace(path, state, false)

export const jump = (n: number | string, state: ?Object, kind: ?string, byIndex: ?boolean) => ({ history }) =>
  history.jump(n, state, kind, byIndex, false)

export const reset = (entries: Array<string | Object>, index: ?number, kind: ?string) => ({ history }) =>
  history.reset(entries, index, kind, false)

export const setState = (state: Object | Function, n: ?(number | string), byIndex: ?boolean) => ({ history }) =>
  history.setState(state, n, byIndex, false)

export const back = (state: ?(Object | Function)) => ({ history }) =>
  history.back(state, false)

export const next = (state: ?(Object | Function)) => ({ history }) =>
  history.next(state, false)
