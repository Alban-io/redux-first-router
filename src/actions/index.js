export { default as redirect } from './redirect'
export { default as notFound } from './notFound'
export { default as addRoutes } from './addRoutes'
export { default as changeBasename } from './changeBasename'

export {
  push,
  replace,
  setState,
  jump,
  back,
  next,
  reset
} from './history'
