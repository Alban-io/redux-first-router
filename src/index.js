export {
  default as connectRoutes,
  go,
  back,
} from './connectRoutes'

export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

export { default as actionToPath } from './pure-utils/actionToPath'
export { default as pathToAction } from './pure-utils/pathToAction'

export type {
  RouteString,
  RouteObject,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
 } from './flow-types'

