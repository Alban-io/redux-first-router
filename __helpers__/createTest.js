import { applyMiddleware, createStore, combineReducers } from 'redux'
import { createRouter } from '../src'

export default async (routesMap, actions, options = {}) => {
  if (actions && !Array.isArray(actions)) {
    options = actions
    actions = undefined
  }

  if (!actions) {
    actions = Object.keys(routesMap).filter(type => !/FIRST|REDIRECTED/.test(type))
  }

  expect(options).toMatchSnapshot('options')

  const routes = {
    FIRST: '/first',
    NEVER_USED_PATHLESS: { // insures pathless routes can co-exist with regular routes
      thunk: jest.fn()
    },
    REDIRECTED: {
      path: '/redirected',
      onComplete: jest.fn(() => 'redirect_complete')
    },
    ...routesMap
  }

  const path = typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
    ? actions.shift()
    : '/first'

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { middleware, reducer, firstRoute, rudy } = createRouter(
    routes,
    options
  )

  const title = (state, action = {}) => {
    return action.payload
      ? action.type + '_' + JSON.stringify(action.payload)
      : action.type
  }
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  if (routesMap.FIRST) {
    expect(store.getState()).toMatchSnapshot('initial_state')
  }

  const firstAction = firstRoute()
  const res = await store.dispatch(firstAction)

  if (routesMap.FIRST) {
    expect(firstAction).toMatchSnapshot('first_action')
    expect(res).toMatchSnapshot('first_response')
    expect(store.getState()).toMatchSnapshot('first_state')
    expect(rudy.history.entries).toMatchSnapshot('first_history_entries')
    expect(rudy.history.index).toMatchSnapshot('first_history_index')
    expect(document.title).toMatchSnapshot('first_title')
  }

  for (const item of actions) {
    let res
    let postfix

    if (typeof item === 'string' && item.charAt(0) === '/') {
      res = await rudy.history.push(item)
      postfix = item
    }
    else {
      const act = typeof item === 'object' ? item : { type: item }
      const { type, params, query, hash, state, snipes } = act
      const action = { type }

      if (params) action.params = params
      if (query) action.query = query
      if (hash) action.hash = hash
      if (state) action.state = state

      res = await store.dispatch(action)
      postfix = '_' + type

      expect(action).toMatchSnapshot('action' + postfix)
    }

    expect(res).toMatchSnapshot('response' + postfix)
    expect(store.getState()).toMatchSnapshot('state' + postfix)
    expect(rudy.history.entries).toMatchSnapshot('history_entries' + postfix)
    expect(rudy.history.index).toMatchSnapshot('history_index' + postfix)
    expect(document.title).toMatchSnapshot('title' + postfix)
  }

  const types = Object.keys(routesMap)

  for (const type of types) {
    const route = routes[type]
    const postfix = '_' + type

    if (route.beforeLeave) {
      expect(route.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave' + postfix)
    }

    if (route.beforeEnter && route.beforeEnter.mock) { // allow redirect shortcut route options to work
      expect(route.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter' + postfix)
    }

    if (route.onLeave) {
      expect(route.onLeave.mock.calls.length).toMatchSnapshot('onLeave' + postfix)
    }

    if (route.onEnter) {
      expect(route.onEnter.mock.calls.length).toMatchSnapshot('onEnter' + postfix)
    }

    if (route.thunk) {
      expect(route.thunk.mock.calls.length).toMatchSnapshot('thunk' + postfix)
    }

    if (route.onComplete) {
      expect(route.onComplete.mock.calls.length).toMatchSnapshot('onComplete' + postfix)
    }

    if (route.onError) {
      expect(route.onError.mock.calls.length).toMatchSnapshot('onError' + postfix)
    }
  }

  if (options.beforeLeave) {
    expect(options.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave_options')
  }

  if (options.beforeEnter) {
    expect(options.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter_options')
  }

  if (options.onLeave) {
    expect(options.onLeave.mock.calls.length).toMatchSnapshot('onLeave_options')
  }

  if (options.onEnter) {
    expect(options.onEnter.mock.calls.length).toMatchSnapshot('onEnter_options')
  }

  if (options.thunk) {
    expect(options.thunk.mock.calls.length).toMatchSnapshot('thunk_options')
  }

  if (options.onComplete) {
    expect(options.onComplete.mock.calls.length).toMatchSnapshot('onComplete_options')
  }

  if (options.onError) {
    expect(options.onError.mock.calls.length).toMatchSnapshot('onError_options')
  }

  return {
    store,
    firstRoute,
    history: rudy.history,
    routes,
    location: () => store.getState().location
  }
}
