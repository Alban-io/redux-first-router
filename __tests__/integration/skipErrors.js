import createTest from '../../__helpers__/createTest'
import { notFound } from '../../src/actions'

createTest('skip route actions with errors', {}, [
  { type: 'REDIRECTED', error: new Error('foo') }
])

