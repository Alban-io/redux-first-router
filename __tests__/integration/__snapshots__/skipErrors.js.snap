// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skip route actions with errors: 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
]
`;

exports[`skip route actions with errors: 1 - history_index 1`] = `0`;

exports[`skip route actions with errors: 1 - response 1`] = `
Object {
  "error": [Error: foo],
  "type": "REDIRECTED",
}
`;

exports[`skip route actions with errors: 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`skip route actions with errors: 1 - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`skip route actions with errors: 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 0,
    "kind": "load",
    "length": 1,
    "params": Object {},
    "pathname": "/first",
    "prev": Object {
      "basename": "",
      "entries": Array [],
      "hash": "",
      "index": -1,
      "kind": "",
      "length": 0,
      "params": Object {},
      "pathname": "",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "type": "",
      "url": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "url": "/first",
  },
  "title": "REDIRECTED",
}
`;

exports[`skip route actions with errors: 1 - title 1`] = `"FIRST"`;
